### PR TITLE
Deployment of the mock ERC677 token contract with Truffle

### PR DESCRIPTION
**Pull request by vkomenda**
- ~The first validator is the owner.~ The owner is different from validators.
- Deployment happens with 0 gas price.

### CODE REVIEW

File: config/node1.toml [hunk 1]
```diff
@@ -30,7 +30,7 @@ password = ["config/password"]
 
 [mining]
 force_sealing = true
-min_gas_price = 1000000000
+min_gas_price = 0
```
Comment (varasev) [inline config/node1.toml]: Arbitrary accounts of the network shouldn't have an ability to use zero gas price. What's the reason of zero on this line?

File: truffle.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      port: 8541,
+      network_id: "*",
+      from: "0xbbcaa8d48289bb1ffcf9808d9aa4b1d215054c78",
+      gas: 8000000,
+      gasPrice: 0
+    }
+  },
+  compilers: {
+    solc: {
+      version: "0.5.2",    // Fetch exact version from solc-bin (default: truffle's version)
```
Comment (varasev) [inline truffle.js]: Let's add `optimizer` and `evmVersion` settings here like in `pos-contracts`: https://github.com/poanetwork/pos-contracts/blob/master/truffle-config.js

File: config/node1.toml (hunk 1 above)
Comment (varasev) [reply config/node1.toml]: To let some account use zero gas price, you can call [`Certifier.certify()`](https://github.com/poanetwork/pos-contracts/blob/924b61525eec1ee38afa4a6ec8d48850343dbadd/contracts/Certifier.sol#L32) and [`TxPermission.addAllowedSender()`](https://github.com/poanetwork/pos-contracts/blob/924b61525eec1ee38afa4a6ec8d48850343dbadd/contracts/TxPermission.sol#L29) by the owner. The owner is added into these contracts by default on genesis, so he can have zero balance and use zero gas price to call those functions.

Comment (vkomenda) [reply config/node1.toml]: Without it I unfortunately get a `truffle migrate` error:
```
Error: Returned values aren't valid, did it run Out of Gas?
```

File: contracts/Hello.sol [hunk 3]
```diff
@@ -0,0 +1,9 @@
+pragma solidity ^0.5.0;
+
+contract Hello {
+    string public message;
+
+    constructor() public {
+        message = "Hello world!";
+    }
+}
```
Comment (afck) [inline contracts/Hello.sol]: Will we remove this contract again once deployment works reliably?

Comment (vkomenda) [reply contracts/Hello.sol]: We could. Or we could keep it and test that a call to `message` returns the hello world string. It's true that we don't need this contract to demonstrate the functionality. It's just a template.

**Comment by vkomenda** [comment]
There can be differences in how deployment succeeds or fails depending on the `aura-pos` revision. @afck, I also had the issue with a single successful deployment once. I had to do `npm run cleanup` and remove `build` to proceed.

**Comment by afck** [comment]
Right, that worked. Should `cleanup` be part of `all`, and should it remove `build`?

**Comment by vkomenda** [comment]
`cleanup` is already in `all`. Although it needs `rm -rf build` to be added.

File: package-lock.json [hunk 4]
```diff
@@ -2801,7 +2801,7 @@
         "lodash": "^4.17.11",
         "oboe": "2.1.4",
         "url-parse": "1.4.4",
-        "websocket": "git://github.com/frozeman/WebSocket-Node.git#browserifyCompatible",
+        "websocket": "git://github.com/frozeman/WebSocket-Node.git#6c72925e3f8aaaea8dc8450f97627e85263999f2",
```
Comment (vkomenda) [inline package-lock.json]: `npm` reverts it back. Why is this manual change needed?

File: config/node1.toml (hunk 1 above)
Comment (vkomenda) [reply config/node1.toml]: @varasev, I didn't try it. It works now, thanks!

File: package-lock.json (hunk 4 above)
Comment (varasev) [reply package-lock.json]: Please never edit `package-lock.json` - it is ofter regenerated by `npm`.

File: contracts/Hello.sol (hunk 3 above)
Comment (vkomenda) [reply contracts/Hello.sol]: I added such test in https://github.com/poanetwork/posdao-test-setup/pull/15/commits/13aa4ea6b6681a9a2e24011d348e39c97b559d86.

**Comment by phahulin** [comment]
Works fine for me.

File: package-lock.json (hunk 4 above)
Comment (vkomenda) [reply package-lock.json]: @varasev, do we need `package-lock.json` to be version-controlled at all? When `npm i` runs, it changes it quite a bit. I'm not sure what difference it makes to have it there in the first place.

Comment (varasev) [reply package-lock.json]: Yes, it should be in the repo: https://stackoverflow.com/questions/44206782/do-i-commit-the-package-lock-json-file-created-by-npm-5

Comment (vkomenda) [reply package-lock.json]: Every time `npm i` is run as part of `npm run all`, this file changes, sometimes dramatically. Therefore I'm not sure about deterministic installs.

**Comment by phahulin** [comment]
@varasev your comment about separating owner from validators should be addressed now, could you check?

**Comment by varasev** [comment]
> @varasev your comment about separating owner from validators should be addressed now, could you check?

@phahulin Yes, I'll check this a bit later.

**Comment by varasev** [comment]
> @varasev your comment about separating owner from validators should be addressed now, could you check?

LGTM

**Comment by vkomenda** [comment]
Tests work faster with the latest changes by @varasev. @DemiMarie, are you still having some issues with this PR?

Can this be merged and tests added in a different PR?

**Comment by phahulin** [comment]
I've also added an initialization for ValidatorSet and ERC677 contracts.
@varasev could you please take a look if it's correct?

**Comment by varasev** [comment]
> I've also added an initialization for ValidatorSet and ERC677 contracts.
> @varasev could you please take a look if it's correct?

ValidatorSetBase is an abstract contract and shouldn't be deployed. Please deploy ValidatorSetAuRa instead.

**Comment by phahulin** [comment]
@varasev :+1: should be addressed now.

File: utils/getContract.js [hunk 5]
```diff
@@ -2,8 +2,8 @@
 const constants = require("./constants");
 
 module.exports = function (contractName, web3) {
-  if (contractName === "ValidatorSetBase") {
-    const VALIDATOR_SET_BASE_ABI = require("../posdao-contracts/build/contracts/ValidatorSetBase").abi;
+  if (contractName === "ValidatorSetAuRa") {
+    const VALIDATOR_SET_BASE_ABI = require("../posdao-contracts/build/contracts/ValidatorSetAuRa").abi;
```
Comment (vkomenda) [inline utils/getContract.js]: Can you drop `BASE` from the name of this variable, e.g., `VALIDATOR_SET_ABI`?

**Comment by varasev** [comment]
Yes, looks good now.

**Comment by phahulin** [comment]
@afck @vkomenda @DemiMarie does this work for you after latest additions? If yes, I think we should merge it.

**Comment by afck** [comment]
Contract deployment (or, more likely, initial funding) fails for me:
```
1_initial_migration.js
======================

   Deploying 'Migrations'
   ----------------------
Error:  *** Deployment Failed ***

"Migrations" could not deploy due to insufficient funds
   * Account:  0x32e4e4c7c5d1cea5db5f9202a9e4d99e56c91a24
   * Balance:  0 wei
   * Message:  Returned error: Insufficient funds. The account you tried to send transaction from does not have enough funds. Required 6600000000000000 and got: 0.
   * Try:
      + Using an adequately funded account
      + If you are using a local Geth node, verify that your node is synced.

    at /home/andreas/git/pos-test-setup/node_modules/truffle/build/webpack:/packages/truffle-deployer/src/deployment.js:364:1
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by vkomenda** [comment]
@phahulin, we should merge it. @DemiMarie?

**Comment by DemiMarie** [comment]
@phahulin @vkomenda I still get insufficient funds errors while deploying.

**Comment by vkomenda** [comment]
@DemiMarie, have you updated `posdao-contracts`? The owner can deploy with 0 gas price.

**Comment by DemiMarie** [comment]
@vkomenda I have.  I still get the error.

**Comment by phahulin** [comment]
@DemiMarie could you post complete log?

**Comment by DemiMarie** [comment]
@phahulin  Nevermind, I was using an old version of the `aura-pos` branch of Parity.

**Pull request merged**